Fix NCCL/Gloo process groups and DDP stream sync bug

mrshenli opened the pull request:
DDP with NCCL backend uses a [worker stream](https://github.com/pytorch/pytorch/blob/d3eb941ed96774efb8d89a0b20c9e49807ea85a7/torch/csrc/distributed/c10d/ddp.cpp#L142) to flatten grand batch
tensors, and passes the flattened tensor to [another stream](https://github.com/pytorch/pytorch/blob/d3eb941ed96774efb8d89a0b20c9e49807ea85a7/torch/lib/c10d/ProcessGroupNCCL.cpp#L379) to
conduct ncclAllReduce. The flattened tensor has to record the
ncclAllReduce stream, otherwise multiple streams might access the
same memory space.

cc @ppwwyyxx

mrshenli committed:
Fix NCCL process group stream sync bug

DDP with NCCL backend uses a worker stream to flatten grand batch
tensors, and passes the flattened tensor to another stream to
conduct ncclAllReduce. The flattened tensor has to record the
ncclAllReduce stream, otherwise multiple streams might access the
same memory space.

mrshenli committed:
add comments

pietern commented:
Wow, this is a tricky one, great catch!

We have some stress/fuzz tests that could (or even should?) have caught this. Can you check why they didn't? It could be because of the allocation pattern being different or not triggering the condition,.

pietern commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -378,6 +378,12 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupNCCL::allreduce(
+    // freed before ncclAllReduce finishes.
+    c10::cuda::CUDACachingAllocator::recordStream(
+      tensors[i].data_ptr(), ncclStream);
+
```
You should move this to _after_ `ncclGroupEnd` to ensure the ordering is correct. This call will record an event to the NCCL stream, and at this location, the NCCL kernel has not been launched yet.

mrshenli replied:
IIUC, this call [adds](https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDACachingAllocator.cpp#L389) the `ncclStream` to a set, and only records an event when the worker stream [attempts to delete the tensor](https://github.com/pytorch/pytorch/blob/master/c10/cuda/CUDACachingAllocator.cpp#L319). In this case, the input `tensors` will not be deleted on the worker stream throughout the `allReduce` invocation. So, it should be OK?

pietern replied:
I see, thanks for the clarification! My understand was that this would record an event, but since it stores the stream the location where this is done doesn't matter.

pietern reviewed (changes requested):
See comment, I think the code block should be moved.

mrshenli commented:
> We have some stress/fuzz tests that could (or even should?) have caught this. Can you check why they didn't? 

May I know which tests are you referring to?

colesbury commented:
I'm concerned that this problem also occurs in a number of other places, including the other ProcessGroupNCCL functions and in ddp.cpp.

I think we need to audit the calls to `cudaStreamWaitEvent` (`CUDAEvent::block`). These often correspond to a use of a tensor in a different stream from which it was allocated.

https://github.com/pytorch/pytorch/blob/20159c3ffe77986ec587b5b0dfa49c6cbbf5a855/torch/lib/c10d/ProcessGroupNCCL.cpp#L70-L80
https://github.com/pytorch/pytorch/blob/20159c3ffe77986ec587b5b0dfa49c6cbbf5a855/torch/lib/c10d/ProcessGroupNCCL.cpp#L365
https://github.com/pytorch/pytorch/blob/20159c3ffe77986ec587b5b0dfa49c6cbbf5a855/torch/lib/c10d/ProcessGroupNCCL.cpp#L412
https://github.com/pytorch/pytorch/blob/20159c3ffe77986ec587b5b0dfa49c6cbbf5a855/torch/lib/c10d/ProcessGroupNCCL.cpp#L460
https://github.com/pytorch/pytorch/blob/20159c3ffe77986ec587b5b0dfa49c6cbbf5a855/torch/lib/c10d/ProcessGroupNCCL.cpp#L529

ddp.cpp
https://github.com/pytorch/pytorch/blob/20159c3ffe77986ec587b5b0dfa49c6cbbf5a855/torch/csrc/distributed/c10d/ddp.cpp#L149
https://github.com/pytorch/pytorch/blob/20159c3ffe77986ec587b5b0dfa49c6cbbf5a855/torch/csrc/distributed/c10d/ddp.cpp#L203

mrshenli commented:
@colesbury thanks, I will screen other occurrences. Regarding `syncStreams`, I think it is necessary, as input tensors are created on the current stream, and then passed to the `ncclStream`. It needs to make sure the communication on the `ncclStream` does not start until the current stream finishes ops on the input tensors.

mrshenli committed:
apply the fix to other ops in DDP and NCCL process group

mrshenli commented:
@colesbury Thanks for pointing out those occurrences. I believe those are all true positives. I have applied changes accordingly.

mrshenli commented:
hmm, changes might also need to be applied to `ProcessGroupGloo`

mrshenli committed:
apply the fix to Gloo process group

pietern commented on torch/csrc/distributed/c10d/ddp.cpp:
```diff
@@ -145,8 +145,18 @@ std::tuple<std::shared_ptr<ProcessGroup::Work>, at::Tensor> queueReduction(
+    // Input `gradsBatch` are created on current streams and used in worker
+    // streams. Hence, they must record worker streams to prevent being
+    // freed before their worker stream ops finish.
+    for (at::Tensor & grad: gradsBatch[devIdx]) {
```
Nit: clang-format this, it will output slightly different code

pietern reviewed (approved):
LGTM (see nit)

mrshenli committed:
edit code format based on clang-format

mrshenli committed:
CUDACachingAllocator keeps memory blocks using storage data_ptr
instead of tensor data_ptr. The tensor data_ptr is sgtorage
data_ptr + offset, where offset could be non-zero if, say, the
tensor is created by slicing another tensor.

mrshenli committed:
fix lint

mrshenli commented:
@pytorchbot retest this please

mrshenli commented:
@pytorchbot retest this please

mrshenli commented:
@pytorchbot retest this please

mrshenli commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

mrshenli commented:
test failures are irrelevant

```
21:36:56 =================================== FAILURES ===================================
21:36:56 _______________ OnnxBackendNodeModelTest.test_where_example_cuda _______________
21:36:56 
21:36:56 args = (<caffe2.python.onnx.tests.onnx_backend_test.OnnxBackendNodeModelTest testMethod=test_where_example_cuda>,)
21:36:56 kwargs = {}
21:36:56 
21:36:56     @unittest.skipIf(  # type: ignore
21:36:56         not self.backend.supports_device(device),
21:36:56         "Backend doesn't support device {}".format(device))
21:36:56     @functools.wraps(test_func)
21:36:56     def device_test_func(*args, **kwargs):  # type: (*Any, **Any) -> Any
21:36:56         try:
21:36:56 >           return test_func(*args, device=device, **kwargs)
21:36:56 
21:36:56 ../.local/lib/python2.7/site-packages/onnx/backend/test/runner/__init__.py:248: 
21:36:56 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
21:36:56 ../.local/lib/python2.7/site-packages/onnx/backend/test/runner/__init__.py:310: in run
21:36:56     outputs = list(prepared_model.run(inputs))
21:36:56 ../.local/lib/python2.7/site-packages/caffe2/python/onnx/backend_rep.py:52: in run
21:36:56     self.workspace.CreateNet(self.predict_net)
21:36:56 ../.local/lib/python2.7/site-packages/caffe2/python/onnx/workspace.py:63: in f
21:36:56     return getattr(workspace, attr)(*args, **kwargs)
21:36:56 ../.local/lib/python2.7/site-packages/caffe2/python/workspace.py:172: in CreateNet
21:36:56     StringifyProto(net), overwrite,
21:36:56 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
21:36:56 
```
